tools/java/org/unicode/cldr/util/XMLSource.java

```diff
@@ -1747,4 +1764,711 @@ public DtdType getDtdType() {
+                currentFullXPath += orderingAttribute();
+            }
+            if (attributes.getLength() > 0) {
+                attributeOrder.clear();
+                for (int i = 0; i < attributes.getLength(); ++i) {
+                    String attribute = attributes.getQName(i);
+                    String value = attributes.getValue(i);
+
+                    if (attribute.equals("cldrVersion")
+                        && (qName.equals("version"))) {
+                        ((SimpleXMLSource) source).setDtdVersionInfo(VersionInfo.getInstance(value));
+                    } else {
+                        putAndFixDeprecatedAttribute(qName, attribute, value);
+                    }
+                }
+                for (Iterator<String> it = attributeOrder.keySet().iterator(); it.hasNext();) {
+                    String attribute = it.next();
+                    String value = attributeOrder.get(attribute);
+                    String both = "[@" + attribute + "=\"" + value + "\"]"; // TODO quote the value??
+                    currentFullXPath += both;
+                    // distinguishing = key, registry, alt, and type (except for the type attribute on the elements
+                    // default and mapping).
+                    // if (isDistinguishing(qName, attribute)) {
+                    // currentXPath += both;
+                    // }
+                }
+            }
+            if (comment != null) {
+                if (currentFullXPath.equals("//ldml") || currentFullXPath.equals("//supplementalData")) {
+                    source.setInitialComment(comment);
+                } else {
+                    source.addComment(currentFullXPath, comment, XPathParts.Comments.CommentType.PREBLOCK);
+                }
+                comment = null;
+            }
+            justPopped = false;
+            lastActiveLeafNode = null;
+            Log.logln(LOG_PROGRESS, "currentFullXPath\t" + currentFullXPath);
+        }
+
+        private String orderingAttribute() {
+            return "[@_q=\"" + (orderedCounter[level]++) + "\"]";
+        }
+
+        private void putAndFixDeprecatedAttribute(String element, String attribute, String value) {
+            if (attribute.equals("draft")) {
+                if (value.equals("true"))
+                    value = "approved";
+                else if (value.equals("false")) value = "unconfirmed";
+            } else if (attribute.equals("type")) {
+                if (changedTypes.contains(element) && isSupplemental < 1) { // measurementSystem for example did not
+                    // change from 'type' to 'choice'.
+                    attribute = "choice";
+                }
+            }
+
+            attributeOrder.put(attribute, value);
+        }
+
+        /**
+         * Types which changed from 'type' to 'choice', but not in supplemental data.
+         */
+        private static Set<String> changedTypes = new HashSet<>(Arrays.asList(new String[] {
```

Comment:
Make it `final` and uppercase (`CHANGED_TYPES = ...`)
And move it at the top of the class, all 3 of them..

Reply:
Agree and will change.